Don't incorporate dependencies' constraint scores multiple times

kgaillot opened the pull request:
The only essential part of this request is the small change in b407688. Everything else is minor refactoring and log message improvement I ran across while investigating it (plus tests).

kgaillot commented:
@wenningerk @gao-yan Can you take a look at this when you have a moment? I'm reasonably confident but I don't want to miss some unintended consequence.

The issue involved group2 non-infinitely colocated with group1. When the first member of group1 was allocated, it would incorporate all the stickiness of group1's members as expected, then it would incorporate the "group2 with group1" constraint by incorporating the constraints from each of group2's members. That would look something like this:

* Incorporate resource 2a stickiness (2a->merge_weights()) at the appropriate factor of 0.002.
* 2a's merge_weights() would then incorporate its own dependencies, namely the internal group colocation "2b with 2a", at full strength. This would recursively include the stickiness for "2c with 2b" and "2d with 2c".
 * Back to group1, it would then incorporate resource 2b's stickiness (2b->merge_weights()) at 0.002. Of course that would incorprate "2c with 2b" and "2d with 2c" at full strength.
* And so on for each member of group2.

Thus 2a's stickiness would wind up counting once, 2b's stickiness would be counted twice, 2c's would be counted three times, etc. This would distort the effect of stickiness vs colocation, and group1 would wind up moving to group2's location (instead of each staying put if stickiness should win, or group2 moving to group1 if colocation should win).

gao-yan commented on lib/pengine/common.c:
```diff
@@ -482,40 +482,73 @@ text2role(const char *role)
+        return CRM_SCORE_INFINITY;
+
+    } else if (score2 >= CRM_SCORE_INFINITY) {
+        crm_trace("+INFINITY + %d = +INFINITY", score1);
```
`%d` first given that it's `score1`?

kgaillot commented:
@gao-yan github won't show me your comment for some reason but I got the email notification:

> In lib/pengine/common.c:
> } else if (score2 >= CRM_SCORE_INFINITY) {
> +        crm_trace("+INFINITY + %d = +INFINITY", score1);
> %d first given that it's score1?

Good catch, fixed and rebased on current master.

gao-yan commented:
> @gao-yan github won't show me your comment for some reason but I got the email notification:

Just moved it into a "review" :-)
